test/cluster/basic_test.go

```diff
@@ -290,3 +291,319 @@ func TestNodeRecoveryLeader(t *testing.T) {
+			}
+			return dataEnv != nil && dataEnv.GetResponse().GetValue() != nil && err == nil
+		}, 30*time.Second, 100*time.Millisecond)
+
+	}
+}
+
+// Scenario:
+// - start 3 servers in a cluster.
+// - wait for one to be the leader.
+// - stop the leader.
+// - wait for new server to be the leader.
+// - the leader submit a tx
+// - repeat until each server was a leader at least once.
+// - make sure each node submitted a valid tx as a leader
+func TestAllNodesGetLeadership(t *testing.T) {
+	dir, err := ioutil.TempDir("", "int-test")
+	require.NoError(t, err)
+
+	nPort, pPort := getPorts(3)
+	setupConfig := &setup.Config{
+		NumberOfServers:     3,
+		TestDirAbsolutePath: dir,
+		BDBBinaryPath:       "../../bin/bdb",
+		CmdTimeout:          10 * time.Second,
+		BaseNodePort:        nPort,
+		BasePeerPort:        pPort,
+		CheckRedirectFunc: func(req *http.Request, via []*http.Request) error {
+			return errors.Errorf("Redirect blocked in test client: url: '%s', referrer: '%s', #via: %d", req.URL, req.Referer(), len(via))
+		},
+	}
+	c, err := setup.NewCluster(setupConfig)
+	require.NoError(t, err)
+	defer c.ShutdownAndCleanup()
+
+	require.NoError(t, c.Start())
+
+	leaders := [3]bool{false, false, false}
+
+	currentLeader := -1
+	require.Eventually(t, func() bool {
+		currentLeader = c.AgreedLeader(t, 0, 1, 2)
+		return currentLeader >= 0
+	}, 30*time.Second, 100*time.Millisecond)
+
+	oldLeader := -1
+	for !leaders[0] || !leaders[1] || !leaders[2] {
+		leaders[currentLeader] = true
+		txID, rcpt, err := c.Servers[currentLeader].WriteDataTx(t, worldstate.DefaultDBName, strconv.Itoa(currentLeader), []byte{uint8(currentLeader)})
+		require.NoError(t, err)
+		require.NotNil(t, rcpt)
+		require.True(t, txID != "")
+		require.True(t, len(rcpt.GetHeader().GetValidationInfo()) > 0)
+		require.Equal(t, types.Flag_VALID, rcpt.Header.ValidationInfo[rcpt.TxIndex].Flag)
+		t.Logf("tx submitted: %s, %+v", txID, rcpt)
+
+		require.NoError(t, c.ShutdownServer(c.Servers[currentLeader]))
+
+		if oldLeader != -1 {
+			require.NoError(t, c.StartServer(c.Servers[oldLeader]))
+		}
+		oldLeader = currentLeader
+		require.Eventually(t, func() bool {
+			currentLeader = c.AgreedLeader(t, (oldLeader+1)%3, (oldLeader+2)%3)
+			return currentLeader >= 0
+		}, 30*time.Second, 100*time.Millisecond)
+	}
+
+	require.NoError(t, c.StartServer(c.Servers[oldLeader]))
+	require.Eventually(t, func() bool {
+		currentLeader = c.AgreedLeader(t, (oldLeader+1)%3, (oldLeader+2)%3)
+		return currentLeader >= 0
+	}, 30*time.Second, 100*time.Millisecond)
+
+	//make sure all 3 nodes are active
+	require.Eventually(t, func() bool {
+		clusterStatusResEnv, err := c.Servers[currentLeader].QueryClusterStatus(t)
+		return err == nil && clusterStatusResEnv != nil && len(clusterStatusResEnv.GetResponse().GetActive()) == 3
+	}, 30*time.Second, 100*time.Millisecond)
+
+	require.Eventually(t, func() bool {
+		return c.AgreedHeight(t, 4, 0, 1, 2)
+	}, 30*time.Second, 100*time.Millisecond)
```

Comment:
Are you sure the height will always be 4? what if the loop runs more times because a node is elected twice before all get elected?
Maybe add a counter to the loop to count how many TXs where actually submitted?

Reply:
Done